### PR TITLE
SWIFT-1260 Disallow embedded null bytes in BSON keys

### PR DESCRIPTION
**Pull request by kmahar**
I did some more re-running of the benchmarks and got the following results:

* JSON to BSON: flat ~5% slower; deep ~7% slower; full  ~10% slower, multi-import ~8% slower
* native to BSON: flat no change, deep ~10% slower, full no change
* runCommand, findOne, insertOne varied across runs, occasionally getting both faster and slower with differences in the range of 1-3%...

For some other benchmarks which do not even touch the new `isValidCString` code, I actually saw both slight speedups and slight slowdowns (like <= 5% change)... I think those can probably just be chalked up to inconsistency on the spawn hosts. 

These numbers do not sound great, but revisiting the scope for the recent BSON performance project, compared to the scope of the performance hits we accepted there (300% slower for JSON to BSON, 25% slower for native to BSON) I don't think these are catastrophic. At the end of the day there is no way around doing this validation so I think we ought to go ahead as-is and if perf comes up as an issue for our users we can revisit and look for other ways to speed up the library.

In summary, what has changed here:
* Synced BSON corpus tests and added new prose tests
* The method we use to encode a C string to a `ByteBuffer`, `writeCString`, now validates there are no embedded null bytes before writing
* As a result of the previous bullet, a number of internal methods now `throw`, hence the number of new `try`s inserted throughout the code
* As a result of the previous bullet, we have to do some new do/catch-ing in places we call those internal methods but are not allowed to throw
* When initializing a `BSONRegularExpression` from JSON, we validate that the pattern and options are valid C strings

### CODE REVIEW

File: Sources/SwiftBSON/BSON.swift [hunk 1]
```diff
@@ -450,7 +450,11 @@ extension BSON: ExpressibleByBooleanLiteral {
 
 extension BSON: ExpressibleByDictionaryLiteral {
     public init(dictionaryLiteral elements: (String, BSON)...) {
-        self = .document(BSONDocument(dictionaryLiteral: elements))
+        do {
```
Comment (kmahar) [inline Sources/SwiftBSON/BSON.swift:453]: the `ExpressibleByDictionaryLiteral` protocol does not allow us to throw from this initializer, so we must crash if the input dictionary literal is invalid

File: Sources/SwiftBSON/BSONDocument.swift [hunk 2]
```diff
@@ -17,11 +17,7 @@ public struct BSONDocument {
 
     internal var storage: BSONDocumentStorage
 
-    internal init(_ elements: [BSON]) {
```
Comment (kmahar) [inline Sources/SwiftBSON/BSONDocument.swift:20]: this initializer was not being used anywhere anymore so I just got rid of it.

File: Sources/SwiftBSON/BSONDocument.swift [hunk 3]
```diff
@@ -185,7 +181,12 @@ public struct BSONDocument {
             BSONDocumentIterator.find(key: key, in: self)?.value
         }
         set {
-            self.set(key: key, to: newValue)
+            // We have no choice but to crash here.
```
Comment (kmahar) [inline Sources/SwiftBSON/BSONDocument.swift:184]: Similarly, we are not allowed to throw from subscript setters/getters. (well, we can as of Swift 5.5, but we support Swift 5.1+ right now). I opened SWIFT-1357 about starting to throw from the subscript in our next major version

File: Package.resolved [hunk 4]
```diff
@@ -33,8 +33,8 @@
         "repositoryURL": "https://github.com/apple/swift-nio",
         "state": {
           "branch": null,
-          "revision": "120acb15c39aa3217e9888e515de160378fbcc1e",
-          "version": "2.18.0"
+          "revision": "6aa9347d9bc5bbfe6a84983aec955c17ffea96ef",
+          "version": "2.33.0"
```
Comment (kmahar) [inline Package.resolved:37]: not strictly related to this PR but NIO required an update to build on Swift 5.5

**Review by patrickfreed — COMMENTED** [review]
one suggestion for a test assertion but otherwise LGTM

File: Tests/SwiftBSONTests/BSONCorpusTests.swift [hunk 5]
```diff
@@ -285,4 +291,89 @@ final class BSONCorpusTests: BSONTestCase {
             }
         }
     }
+
+    /// Tests from the "Prose Tests" section of the test plan.
+    func testBSONCorpusProse() throws {
+        // 1. Prohibit null bytes in null-terminated strings when encoding BSON
+        // Note we can't write these tests using the dictionary literal syntax since all of these cases trigger
+        // fatalErrors with that API.
```
[the remark this answers is not in the document]
Comment (kmahar) [reply Tests/SwiftBSONTests/BSONCorpusTests.swift:299]: nice, I added additional expectations that test the public API explicitly

[the remark this answers is not in the document]
Comment (patrickfreed) [reply Tests/SwiftBSONTests/BSONCorpusTests.swift]: Unfortunately it looks like using that in combination with the leak checker crashes the compiler per the evergreen output. Maybe we just have to go back to the original test after all :disappointed:

[the remark this answers is not in the document]
Comment (kmahar) [reply Tests/SwiftBSONTests/BSONCorpusTests.swift]: ugh. yeah I guess so. will revert my last two commits

**Review by patrickfreed — APPROVED** [review]
LGTM!

**Review by isabelatkinson — APPROVED** [review]
